engineio/context.go

```diff
@@ -0,0 +1,33 @@
+package engineio
+
+import (
+	"context"
+	"io"
+	"net/http"
+
+	"github.com/googollee/go-socket.io/engineio/packet"
+)
+
+// Context provides info for a HTTP request.
+type Context struct {
+	// Standard context
+	context.Context
+
+	// The instance of current session.
+	Session Session
+
+	// The http.Request of current HTTP request.
+	// It may be different when using polling transport.
+	Request *http.Request
+
+	// The current packet infomation.
+	Packet Packet
+}
+
+type Packet struct {
```

Comment:
Why are you want join `packet.Type` and `Body` by nested type?

Reply:
Because they are not something of context, they are requests. I'm also thinking about splitting it from context and change the signature to `OnMessage(ctx *Context, request Packet)`. `With()` is in the same way.

Reply:
> I'm also thinking about splitting it from context

It's good idea. i suggest usage the next type:

```go
type Packet struct {
io.Reader
Type packet.Type
}
```
The next code allow use as `ctx.Packet.Read()` This is simple usage ctx by client

Reply:
Done.